### PR TITLE
internal/testutil: manage GCP resource creation and cleanup

### PR DESCRIPTION
**Pull request by rakyll**

### CODE REVIEW

**Comment by broady** [comment]
```
Step #2: storage/buckets/main_test.go:18: undefined: testutil.Namespace
Step #2: FAIL	github.com/GoogleCloudPlatform/golang-samples/storage/buckets [build failed]
```

File: bigquery/snippets/snippet_test.go [hunk 1]
```diff
@@ -9,16 +9,14 @@ import (
...
 
 	"cloud.google.com/go/bigquery"
-	"cloud.google.com/go/storage"
 	"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
 	"golang.org/x/net/context"
-	"golang.org/x/oauth2/google"
-	rawbq "google.golang.org/api/bigquery/v2"
 )
 
+const ns = "snippets"
```
Comment (broady) [inline bigquery/snippets/snippet_test.go]: this needs to be unique across all tests

`const ns = "bigquery-snippets`

File: internal/testutil/resources.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+// NextObject will return a new unique object name for the given namespace.
+func NextObject(namespace string) string {
+	atomic.AddUint64(&idCounter, 1)
+	return fmt.Sprintf("%s_%s_object%d_%d", appPrefix, namespace, idCounter, time.Now().Unix())
+}
+
+// CleanupDatasets deletes the given datasets and all expired
+// datasets belongs to the tests.
+func CleanupDatasets(t *testing.T, name ...string) {
```
Comment (broady) [inline internal/testutil/resources.go]: Why doesn't it take a namespace?

File: internal/testutil/resources.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+	idCounter uint64
+	expiry    = time.Hour
+)
+
+const appPrefix = "golang_samples"
+
+// NextDataset returns the next unique dataset ID generated for the given namespace.
+func NextDataset(namespace string) string {
+	atomic.AddUint64(&idCounter, 1)
+	return fmt.Sprintf("%s_%s_dataset%d_%d", appPrefix, namespace, idCounter, time.Now().Unix())
```
Comment (broady) [inline internal/testutil/resources.go]: This is racy.
```
id := atomic.AddUint64(&idCounter, 1)
return fmt.Sprintf("%s_%s_dataset%d_%d", appPrefix, namespace, id, time.Now().Unix())
```

File: internal/testutil/resources.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+var (
+	idCounter uint64
+	expiry    = time.Hour
+)
+
+const appPrefix = "golang_samples"
+
+// NextDataset returns the next unique dataset ID generated for the given namespace.
+func NextDataset(namespace string) string {
+	atomic.AddUint64(&idCounter, 1)
```
Comment (jba) [inline internal/testutil/resources.go]: I think mutexes are fine unless performance is really an issue.

File: internal/testutil/resources.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+	return fmt.Sprintf("%s_%s_dataset%d_%d", appPrefix, namespace, idCounter, time.Now().Unix())
+}
+
+// NextTable returns the next unique table ID generated for the given namespace
+func NextTable(namespace string) string {
```
Comment (jba) [inline internal/testutil/resources.go]: It seems like these could all be one function.

Or at least they should call one more general, unexported function.

File: internal/testutil/resources.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+	hc, err := google.DefaultClient(ctx, rawbq.CloudPlatformScope)
+	if err != nil {
+		t.Errorf("DefaultClient: %v", err)
+	}
+	s, err := rawbq.New(hc)
+	if err != nil {
+		t.Errorf("bigquery.New: %v", err)
+	}
+	call := s.Datasets.Delete(tc.ProjectID, datasetID)
```
Comment (jba) [inline internal/testutil/resources.go]: This is now available in cloud.google.com/go/bigquery.

**Comment by broady** [comment]
I think this PR is obsolete, in light of the testing project pool (that is, no need to think about concurrent builds). Resource IDs can stay the same between test runs, as long as they incorporate the project ID in their name.

**Pull request closed without merging**